sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,572 @@
+{
+  auto httpVersionEnd = httpVersion.data() + httpVersion.size();
+
+  // set response code, http version and reason phrase (i.e. HTTP/1.1 200 OK)
+  auto majorVersionStart
+      = httpVersion.data() + 5; // HTTP = 4, / = 1, moving to 5th place for version
+  auto majorVersionEnd = std::find(majorVersionStart, httpVersionEnd, '.');
+  auto majorVersionInt = std::stoi(std::string(majorVersionStart, majorVersionEnd));
+
+  auto minorVersionStart = majorVersionEnd + 1; // start of minor version
+  auto minorVersionInt = std::stoi(std::string(minorVersionStart, httpVersionEnd));
+
+  *majorVersion = (uint16_t)majorVersionInt;
+  *minorVersion = (uint16_t)minorVersionInt;
+}
+
+void CleanupHandles(HINTERNET sessionHandle, HINTERNET connectionHandle, HINTERNET requestHandle)
+{
+
+  if (requestHandle)
+  {
+    WinHttpCloseHandle(requestHandle);
+  }
+
+  if (connectionHandle)
+  {
+    WinHttpCloseHandle(connectionHandle);
+  }
+
+  if (sessionHandle)
+  {
+    WinHttpCloseHandle(sessionHandle);
+  }
+}
+
+void CleanupHandlesAndThrow(
+    std::string exceptionMessage,
+    HINTERNET sessionHandle = NULL,
+    HINTERNET connectionHandle = NULL,
+    HINTERNET requestHandle = NULL)
+{
+  DWORD error = GetLastError();
+
+  CleanupHandles(sessionHandle, connectionHandle, requestHandle);
+
+  throw Azure::Core::Http::TransportException(
+      exceptionMessage + " Error Code: " + std::to_string(error) + ".");
+}
+
+std::unique_ptr<RawResponse> WinHttpTransport::Send(Context const& context, Request& request)
+{
+  // TODO: Need to test if context has been canceled and cleanup, can't call ThrowIfCanceled.
+  (void)(context);
+
+  // Use WinHttpOpen to obtain a session handle.
+  // The dwFlags is set to 0 - all WinHTTP functions are performed synchronously.
+  // TODO: Use specific user-agent or application name.
+  HINTERNET sessionHandle = WinHttpOpen(
+      L"WinHTTP Azure SDK",
+      WINHTTP_ACCESS_TYPE_NO_PROXY,
+      WINHTTP_NO_PROXY_NAME,
+      WINHTTP_NO_PROXY_BYPASS,
+      0);
+
+  if (!sessionHandle)
+  {
+    // Errors include:
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_NOT_ENOUGH_MEMORY
+    CleanupHandlesAndThrow("Error while getting a session handle.");
+  }
+
+  // TODO: Get port from Url
+  // Specify an HTTP server.
+  // Uses port 80 for HTTP and port 443 for HTTPS.
+  // This function always operates synchronously.
+  HINTERNET connectionHandle = WinHttpConnect(
+      sessionHandle,
+      StringToWideString(request.GetUrl().GetHost()).c_str(),
+      INTERNET_DEFAULT_PORT,
+      0);
+
+  if (!connectionHandle)
+  {
+    // Errors include:
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_INVALID_URL
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+    // ERROR_WINHTTP_SHUTDOWN
+    // ERROR_NOT_ENOUGH_MEMORY
+    CleanupHandlesAndThrow("Error while getting a connection handle.", sessionHandle);
```

Comment:
HandleManager isn't really a BodyStream and the lifetimes of the three handles are different. But I will play around with your suggestion.

Reply:
you are right.
I think the session handle should not be copied to the bodyStream, it should live only in the transportAdapter.
The bodyStream doesn't need the session handle to keep reading from network.
I think it only needs the request Handle and the connection handle (just to clean it once we finish).

The HandleManager is not initially a bodyStream, but it is meant to become one.
You are currently taking part of the HandleManager to create a bodyStream,  the other approach is that it is since the beginning a bodyStream and once it is returned as such, it loses all the things that are not part of it. Then you don't need 2 types, but only one